rust-src/concordium_base/src/id/sigma_protocols/com_ineq.rs

```diff
@@ -0,0 +1,174 @@
+//! The module provides the implementation of the sigma protocol for "proof of
+//! inequality for committed value and public value". This protocol enables one
+//! to prove that a committed value is not equal to a public one , without
+//! revealing the value.
+
+use super::{
+    com_mult::{ComMult, ComMultSecret, Witness as ComMultWitness},
+    common::{prove as sigma_prove, verify as sigma_verify, SigmaProof},
+};
+use crate::{
+    common::*,
+    curve_arithmetic::Curve,
+    pedersen_commitment::{Commitment, CommitmentKey, Randomness, Value},
+    random_oracle::RandomOracle,
+};
+use ff::Field;
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
+pub struct Witness<C: Curve> {
+    // The witness consists of a com_mult witness and a commitment.
+    com_mult_witness: SigmaProof<ComMultWitness<C>>,
+    aux_com:          Commitment<C>,
+}
+
+/// Function for proving that a committed value `value` is different from
+///  `pub_value`. The parameters are
+/// - `com_key` - the commitment key used to commit to the value
+/// - `value` - the value inside the commitment
+/// - `value_tilde` - the randomness used to commit
+/// - `pub_value` - the public value claimed to be different from `value`
+/// - `csprng` - a cryptographically secure random number generator
+
+pub fn prove_com_ineq<R: rand::Rng, C: Curve>(
```

Comment:
I am not so keen on this interface. For other sigma protocols we have the uniform

- define a struct for various parts (public, secret)
- implement a SigmaProtocol trait for it

This deviates from it which means we won't be able to compose them, like we can for other sigma protocols.

Reply:
That's true. The main reason is that non-equality

Reply:
That's true. The main reason is that the structure of inequality proofs is a bit different from that of other Sigma protocols we have. In particular, the commitment to the inverse of (value - pub_value) in inequality proofs is not known to the verifier in advance and should come as part of the prover's third message. This makes inequality proofs somehow incompatible with other sigma protocols and probably not easily composable as you said.